M2/Macaulay2/packages/Hadamard.m2

```diff
@@ -0,0 +1,437 @@
+newPackage(
+    "Hadamard",
+    Version => "0.1",
+    Date => "November 2020",
+    Authors => {
+
+	{Name => "Iman Bahmani Jafarloo",
+	    Email => "ibahmani89@gmail.com",
+	    HomePage => "http://calvino.polito.it/~imanbj"
+	    }
+	}, -- TODO
+    Headline => "A package for the Hadamard products of projective subvarieties",
+    AuxiliaryFiles => false,
+    DebuggingMode => false,
+    Reload => false,
+    PackageExports => {"Points"},
+    Keywords => {"Commutative Algebra"}
+    )
+export {
+    -- types
+    "Point",
+    -- methods
+    "point",
+    "hadamardProduct",
+    "hadamardPower",
+    "idealOfProjectivePoints"
+    }
+
+--defining a new type of objects: points
+
+Point = new Type of BasicList
+
+point=method()
+point(VisibleList):=(P)->(
+    if all(P, i->i==0) then return "error: all entries are zero" else
+    if any(P,i->i===null) then return "error: nuul entries are not allowed"
+    else new Point from P)
+
+
+Point * Point:=(p,q)->(
+    if #p =!= #q then error("points should be in a same projective space") else
+    pp:=apply(p,q,times);
+    if all(pp, i->i==0) then error("product of points has no nonzero coordinate")
+    else return pp
+    )
+
+Point == Point := (p,q)->(
+    rank pointsToMatrix({p,q}) == 1
+    )
+
+---Hadamard product of two points---
+hadProdPoints = method()	 
+hadProdPoints(Point,Point):=(p,q)->(
+     p * q
+    )
+
+hadProdPoints(List,List):=(p,q)->(
+ point p * point q
+    )
+
+hadProdPoints(Array,Array):=(p,q)->(
+   point p * point q
+    )
+
+-- Hadamrd product of two varieties
+hadProdOfVariety = method()
+hadProdOfVariety (Ideal, Ideal):= (I,J) -> (
+    newy := symbol newy;
+    newz := symbol newz;
+    varI:= gens ring I;
+    varJ:= gens ring J;
+    CRI:=coefficientRing ring I;
+    CRJ:=coefficientRing ring J;
+    RYI:=CRI[newy_0.. newy_(# varI -1)];
+    RZJ:=CRI[newz_0..newz_(# varJ -1)];
+    IY:=sub(I,apply(varI,gens RYI,(a,b)->a=>b));
+    JZ:=sub(J,apply(varJ,gens RZJ,(a,b)->a=>b));
+    TensorRingIJ:=RYI/IY ** RZJ/JZ;
+    use TensorRingIJ;
+    Projvars:=apply(#(gens ring I),i->newy_i * newz_i);
+    hadMap:=map(TensorRingIJ,ring I, Projvars);
+    ker hadMap
+    )
+
+
+-------Hadmard product of two subsets of points on two varieties---------
+hadProdListsOfPoints = method()
+hadProdListsOfPoints(List,List) :=(X,Y)->(
+     convert:= obj -> if not instance(obj, Point) then point(obj) else obj;
+     newX:=apply(X,convert);
+     newY:=apply(Y,convert);
+     return delete(null, for I in toList(set(newX) ** set(newY)) list (
+	 try(I_0 * I_1) then (I_0 * I_1) else null
+	 ))
+     )
+
+
+pointsToMatrix=method()
+pointsToMatrix(List):= (PTM) ->( matrix apply(PTM, toList))
+
+---Hadamard powers of varieties------------
+
+hadamardPower = method()
+hadamardPower(Ideal,ZZ):=(I,r)->(
+    if r<1 then return error("the second argument should be positve integre >=1 ");
```

Comment:
After an error, execution is interrupted, so the return on this line is never executed. Just remove it.